### ISSUE TITLE
NullReferenceException for multiple implicit stream subscriptions

### ISSUE THREAD

**Issue by charleszipp** [dotnet/orleans#3212]
I am trying to create an actor that subscribes to multiple streams on the same stream namespace. This is done to facilitate both fire-and-forget style calls but, more importantly, to remove the temporal dependency between this actor and other actors in different solutions.

When i produce the CommandDispatched event, the actor is activated as expected. However, i want to also ensure a subscription is setup for other events (CommandStarted, CommandCompleted, etc). When it hits the second subscription for CommandStarted, a NullReferenceException is thrown.

At the time the exception is thrown, there are not actually any CommandStarted events on the stream. However, i need some way to subscribe to this stream before its events are produced. Do i need to do an explicit subscription here? Possibly from the CommandDispatched handler?

```csharp
[ImplicitStreamSubscription("Routes.Activities")]
public class ActivityMonitor : JournaledGrain<ActivityState, IEvent>, IActivityMonitor
{
	protected Task OnObserveCommandDispatched(CommandDispatched @event, StreamSequenceToken token = null)
    {
        //handle event
        return Task.CompletedTask;
    }

    protected Task OnObserveCommandStarted(CommandStarted @event, StreamSequenceToken token = null)
    {
        //handle event
        return Task.CompletedTask;
    }

    public override async Task OnActivateAsync()
    {
        await GetStreamProvider("activities").GetStream<CommandDispatched>(this.GetPrimaryKey(), "Routes.Activities").SubscribeAsync(OnObserveCommandDispatched);
        //this throws exception object null reference exception
        await GetStreamProvider("activities").GetStream<CommandStarted>(this.GetPrimaryKey(), "Routes.Activities").SubscribeAsync(OnObserveCommandStarted);
    }
}

public class Program
{
	private static IClientFactory _clients = new ServiceFabricClusterClientFactory();
	private static void Main(string[] args)
	{
		var client = _clients.Get<IActivityMonitor>().Result;
		var command = new CreateEngagementArgs(id, "Test 123", firmGuid);
        var @event = new EngagementCreated(id, "Test 123", firmGuid);

        client.GetStreamProvider("activities").GetStream<CommandDispatched>(activityId, "Routes.Activities").OnNextAsync(
            new CommandDispatched(activityId, "Routes.Activities", Newtonsoft.Json.JsonConvert.SerializeObject(command), DateTime.UtcNow)
            ).Wait();

        //todo: produce CommandStarted...
	}
}
```

```
Audit.Sagas.Actors.exe Error: 0 : [2017-07-11 10:44:17.791 GMT    22	ERROR  	100513	Catalog	127.0.0.1:30023]	!!!!!!!!!! Error calling grain's OnActivateAsync() method - Grain type = Audit.Sagas.Actors.ActivityMonitor Activation = [Activation: S127.0.0.1:30023:237465823*grn/282D34D6/8d50069f@14f196be #GrainType=Audit.Sagas.Actors.ActivityMonitor Placement=RandomPlacement State=Activating]	
Exc level 0: System.NullReferenceException: Object reference not set to an instance of an object.
   at Orleans.Streams.AsyncObservableExtensions.SubscribeAsync[T](IAsyncObservable`1 obs, Func`3 onNextAsync, Func`2 onErrorAsync, Func`1 onCompletedAsync)
   at Orleans.Streams.AsyncObservableExtensions.SubscribeAsync[T](IAsyncObservable`1 obs, Func`3 onNextAsync)
   at Audit.Sagas.Actors.ActivityMonitor.<OnActivateAsync>d__7.MoveNext() in C:\Code\Test\Audit\Audit\Audit.Sagas.Actors\ActivityMonitor.cs:line 370
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.Runtime.Catalog.<CallGrainActivate>d__87.MoveNext()
Audit.Sagas.Actors.exe Warning: 0 : [2017-07-11 10:44:17.796 GMT    25	WARNING	100534	Catalog	127.0.0.1:30023]	Failed to InvokeActivate for [Activation: S127.0.0.1:30023:237465823*grn/282D34D6/8d50069f@14f196be #GrainType=Audit.Sagas.Actors.ActivityMonitor Placement=RandomPlacement State=Invalid].	
Exc level 0: System.NullReferenceException: Object reference not set to an instance of an object.
   at Orleans.Streams.AsyncObservableExtensions.SubscribeAsync[T](IAsyncObservable`1 obs, Func`3 onNextAsync, Func`2 onErrorAsync, Func`1 onCompletedAsync)
   at Orleans.Streams.AsyncObservableExtensions.SubscribeAsync[T](IAsyncObservable`1 obs, Func`3 onNextAsync)
   at Audit.Sagas.Actors.ActivityMonitor.<OnActivateAsync>d__7.MoveNext() in C:\Code\Test\Audit\Audit\Audit.Sagas.Actors\ActivityMonitor.cs:line 370
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.Runtime.Catalog.<CallGrainActivate>d__87.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.Runtime.Scheduler.SchedulerExtensions.<>c__DisplayClass1_0.<<QueueTask>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.Runtime.Catalog.<InitActivation>d__66.MoveNext()
```

**Comment by jason-bragg** [comment]
The null reference error you're seeing should not be happening, so it's something we'll look into, and I thank you for bringing it to our attention.

Having said that, the pattern your using is not one we support.  The 'type' is not considered part of the stream identity so GetStream<CommandDispatched>(this.GetPrimaryKey(), "Routes.Activities") and GetStream<CommandStarted>(this.GetPrimaryKey(), "Routes.Activities") both subscribe to the same stream, and any events encountered on those streams other than the specified type will generate warnings and/or errors.

For what it looks like you're trying to do, the usual pattern is to have a base class that all events inherit from and use that for the stream type, then dispatch to specific types in the handler.  This will, however, put all of the events in a single stream, which may not be what you want.  Alternatively you can use different namespaces for each type, and have a single grain be implicitly subscribed to multiple namespaces.

**Comment by xiazen** [comment]
I reproduced the bug or weird error you are seeing. The cause conforms to @jason-bragg 's comments

Like Jason said, type is not part of steam identity, but stream itself is typed. So in the first 
`GetStreamProvider("activities").GetStream<CommandDispatched>(this.GetPrimaryKey(), "Routes.Activities").`, it created a `IAsyncStream<CommandDispatched> `in streaming infra. Then In the second subscribe async, 
the code `GetStreamProvider("activities").GetStream<CommandStarted>(this.GetPrimaryKey(), "Routes.Activities")` trying to get the stream with the same identity will return null, which is what triggered the NullReferenceException, because it is casting a `IAsyncStream<CommandDispatched>` to a 
`IAsyncStream<CommandStarted> `

But whether it is a good user experience is still debatable. I can see how it can be confusing for user to figure out the correct way to use SubscribeAsync. Maybe adding more documentation around the API can help .

**Comment by charleszipp** [comment]
Thanks for the helpful info @xiazen and @jason-bragg!

It appears this was a misunderstanding on my part. Jason's statement was exactly what i was thinking where type is part of the stream identity. Now hearing this feedback, i think it not being part of the stream identity actually makes more sense to me. The reason is because actually what I believe i should be doing is keeping all of these events on the same stream. I want the events to stay ordered Dispatched > Started > Completed, etc. Since im using the event hub stream provider, keeping these on the same stream should ensure they are ordered and replayable in that order. If they were in different streams and i replayed the streams, i believe they could then happen in a different order than what actually occurred.

With that said, ill give the abstract event method a go and see if i can use that to keep them on the same stream. 

The only suggestion i would make in terms of changes would be 1) throwing a more descriptive error that may help someone figure out this is wrong on their own and 2) adding something to the stream documentation (if not there already) about the different methods for subscribing to multiple events as Jason described above. 

Thanks again for the quick, thoughtful, and helpful responses!

**Comment by xiazen** [comment]
@charleszipp glad to help! the two suggestions you had there is also among the options me and Jason has discussed. Thanks for the feedback !

**Comment by charleszipp** [comment]
Just wanted to fyi, that @jason-bragg's suggestion of using an abstract/interface event to put multiple events on a single stream worked for me. I was able to do what i wanted using this approach. Only additional thing i had to add was a mediator to route the event to the appropriate handler

**Comment by sergeybykov** [comment]
We should throw an exception with an explicit error message instead of returning a null stream in case of stream type mismatch.

**Comment by xiazen** [comment]
yes agreed.

**Comment by xiazen** [comment]
Submitted a fix in #3301

**Comment by xiazen** [comment]
close this since the fix is submitted
